src/Connection/RelayFactory.php

```diff
@@ -0,0 +1,147 @@
+     */
+    public function create($parameters): NodeConnectionInterface
+    {
+        $this->assertExtensions();
+
+        if (!$parameters instanceof ParametersInterface) {
+            $parameters = $this->createParameters($parameters);
+        }
+
+        $scheme = $parameters->scheme;
+
+        if (!isset($this->schemes[$scheme])) {
+            throw new InvalidArgumentException("Unknown connection scheme: '$scheme'.");
+        }
+
+        $initializer = $this->schemes[$scheme];
+        $client = $this->createClient();
+
+        $connection = new $initializer($parameters, $client);
+
+        $this->prepareConnection($connection);
+
+        return $connection;
+    }
+
+    /**
+     * Checks if the Relay extension is loaded in PHP.
+     */
+    private function assertExtensions()
+    {
+        if (!extension_loaded('relay')) {
+            throw new NotSupportedException(
+                'The "relay" extension is required by this connection backend.'
+            );
+        }
+    }
+
+    /**
+     * Creates a new instance of the client.
+     *
+     * @return Relay
+     */
+    private function createClient()
+    {
+        $client = new Relay();
+
+        // throw when errors occur and return `null` for non-existent keys
+        $client->setOption(Relay::OPT_PHPREDIS_COMPATIBILITY, false);
+
+        // use reply literals
+        $client->setOption(Relay::OPT_REPLY_LITERAL, true);
+
+        // disable Relay's command/connection retry
+        $client->setOption(Relay::OPT_MAX_RETRIES, 0);
+
+        // whether to use in-memory caching
+        $client->setOption(Relay::OPT_USE_CACHE, $this->parameters->cache ?? true);
+
+        // set data serializer
+        $client->setOption(Relay::OPT_SERIALIZER, constant(sprintf(
+            '%s::SERIALIZER_%s',
+            Relay::class,
+            strtoupper($this->parameters->serializer ?? 'none')
+        )));
+
+        // set data compression algorithm
+        $client->setOption(Relay::OPT_COMPRESSION, constant(sprintf(
+            '%s::COMPRESSION_%s',
+            Relay::class,
+            strtoupper($this->parameters->compression ?? 'none')
+        )));
+
+        return $client;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function prepareConnection(NodeConnectionInterface $connection)
+    {
+        $parameters = $connection->getParameters();
+
+        if (isset($parameters->password) && strlen($parameters->password)) {
+            $cmdAuthArgs = isset($parameters->username) && strlen($parameters->username)
+                ? [$parameters->username, $parameters->password]
+                : [$parameters->password];
+
+            $connection->addConnectCommand(
+                new RawCommand('AUTH', $cmdAuthArgs)
+            );
+        }
+
+        if (isset($parameters->database) && strlen($parameters->database)) {
+            $connection->addConnectCommand(
+                new RawCommand('SELECT', [$parameters->database])
```

Comment:
When using Relay `auth()` and `select()` are special methods that should be executed, without using a raw command. 

Correct me if I'm wrong @michael-grunder.

Reply:
I think `select` can be used via `rawCommand` but `auth` should not be really.

Reply:
Ah. I thought it switches in the internal database for caching or something similar.

Reply:
@tillkruss But this commands will be executed as `auth()` and `select()`, because `commandId` matching `$atypicalCommands` values.

https://github.com/predis/predis/blob/feature/relay-improvements/src/Connection/RelayConnection.php#L70
https://github.com/predis/predis/blob/feature/relay-improvements/src/Connection/RelayConnection.php#L176

Reply:
Right, glad you confirmed.